### DISCUSSION TITLE
我想在云服务器上挂脚本，有什么可以在服务器上用的模拟器吗

### DISCUSSION THREAD

**Discussion by Ak1zukiKanna** [MaaAssistantArknights/MaaAssistantArknights#3241] [Q&A] [answered]
服务器小白，在此之前不知道HYPER-V不兼容的问题··

**Comment by horror-proton** [comment]
Linux 服务器的话 @aa889788 应该有相关经验

**Reply by aa889788** [reply]
我是在Linux上用redroid的，不过redroid跟跑了个android x86差不多，一般云服务器要是没显卡加速跑方舟估计够呛

**Reply by length18cm** [reply]
> 我是在Linux上用redroid的，不过redroid跟跑了个android x86差不多，一般云服务器要是没显卡加速跑方舟估计够呛

大佬，请问在linux上跑，有专门的gui吗？

**Reply by horror-proton** [reply]
MAA 暂时没 Linux gui, 可以选择用 Python 脚本调用, 远程操作安卓可以用 scrcpy

**Reply by bigqiao** [reply]
都在Linux上跑了，应该要用能力用Python自己集成了吧，摊手

**Reply by length18cm** [reply]
啊这，我真是小白，只会按教程按部就班的来

**Reply by horror-proton** [reply]
直接把 linux 包下载到合适的位置, 然后执行 Python/sample.py 应该就行吧 (脚本里面的文件路径和 adb 地址可能需要调整一下), 可以根据文档修改里面的任务参数

**Reply by length18cm** [reply]
谢谢大佬，我去试试

**Comment by horror-proton** [comment, marked as answer]
我在我的 Linux 服务器上试了一下 redroid, 感觉还是可以跑的
硬件是一个腾讯云的一个 2 核轻量应用服务器, 没显卡, **跑方舟最好保证至少 4G 内存**, (我这里物理内存只有 2G, 所以又加了 2G 的 swapfile)

-----
- 如果希望使用 release 中预编译的 MAA, 最好保证环境中的 libc 和 libstdc++ 等基础库是较为新的版本, 不然可能会遇到一些奇奇怪怪的 ABI 版本问题, (release 中的二进制版本是在 Ubuntu-22.04 中编译的, 很大概率无法直接在比较老的发行版上运行)
这个问题在 pr3346 合并后可能可以缓解
- 未来 Arch Linux 用户可以尝试参考 [maa-assistant-arknights](
https://aur.archlinux.org/packages/maa-assistant-arknights), [maa-assistant-arknights-git](https://github.com/horror-proton/PKGBUILDs/tree/master/maa-assistant-arknights-git) 打包
-----
- redroid 需要一些内核功能如 binderfs, 在通常不会默认被打开, 需要参考 [redroid 文档](https://github.com/remote-android/redroid-doc/tree/master/deploy) 根据自己使用的发行版修改 Linux 内核参数或者更换内核. 
- 按照 redroid 文档中的参数使用 `docker run` 运行 redroid, ~~podman 我试了下没成功,~~ `docker ps` 观察是否正在运行
- 尝试连接 adb connect localhost:5555, 如果无法连接, 使用 `docker exec -it redroid?? sh` 进入安卓的 shell, 然后用 `logcat` 和 `ps` 等工具观察 Android 系统是否正常运行, adbd 是否正常启动, `killall adbd` 使 adbd 重启并观察 logcat 输出.
- 可以使用 [scrcpy](https://github.com/Genymobile/scrcpy) 远程控制安卓, 比如在 Windows 上
  ```pwsh
  cd scrcpy-win64-v1.25
  ./adb.exe connect 服务器IP:5555
  ./scrcpy.exe
  ```
  或者使用界面更友好的 [scrcpy-plus](https://github.com/Frontesque/scrcpy-plus), [QtScrcpy](https://github.com/barry-ran/QtScrcpy)
- 下载明日方舟 apk 并安装, 如
  ```bash
  cd ~/data11/local/tmp # ~/data11 是容器的共用目录, 取决于 docker run 的参数
  sudo wget --content-disposition https://ak.hypergryph.com/downloads/android_lastest
  docker exec redroid11 sh -c 'pm install /data/local/tmp/*.apk'
  ```
-----
- 根据需要编辑 `Python/sample.py` 中的内容
- MAA 默认的 minitouch 方案可能无法使用, 取消 maatouch 一行前的注释以使用 maatouch
- 运行时方舟可能帧率只有个位数, 可以尝试适当调大 MAA 的 resource/config.json 中的 `"taskDelay"` 参数.

![image](https://user-images.githubusercontent.com/107091537/219077419-49b2ed90-2797-42a3-a359-4d62acd1de12.png)

**Reply by Aoi-cn** [reply]
太牛逼了。。

**Comment by qaz1qazlol2** [comment]
我把maa和任务计划程序加上我自己写的粗略的阿里云云手机sdk配合使用来达到这个效果
